tsdb/store_test.go

```diff
@@ -262,6 +262,70 @@ func TestStore_BadShard(t *testing.T) {
 	}
 }
 
+func TestStore_BadShardClear(t *testing.T) {
+	const errStr = "a shard open error"
+	indexes := tsdb.RegisteredIndexes()
+	for _, idx := range indexes {
+		func() {
+			s := MustOpenStore(t, idx)
+			defer require.NoErrorf(t, s.Close(), "closing store with index type: %s", idx)
+
+			sh := tsdb.NewTempShard(idx)
+			shId := sh.ID()
+			err := s.OpenShard(sh.Shard, false)
+			require.NoError(t, err, "opening temp shard")
+			defer require.NoError(t, sh.Close(), "closing temporary shard")
+
+			expErr := errors.New(errStr)
+			s.SetShardOpenErrorForTest(sh.ID(), expErr)
+			err2 := s.OpenShard(sh.Shard, false)
+			require.Error(t, err2, "no error opening bad shard")
+			require.ErrorIs(t, err2, tsdb.ErrPreviousShardFail{})
+			require.EqualError(t, err2, fmt.Errorf("not attempting to open shard %d; opening shard previously failed with: %w", shId, expErr).Error())
+
+			require.Equal(t, 1, len(s.Store.GetBadShardList()))
+
+			badShards := s.ClearBadShardList()
+			require.Len(t, badShards, 1)
+
+			// Check that bad shard list has been cleared
+			require.Empty(t, s.Store.GetBadShardList())
+
+			s.SetShardOpenErrorForTest(sh.ID(), expErr)
+			err2 = s.OpenShard(sh.Shard, false)
+			require.Error(t, err2, "no error opening bad shard")
+			require.ErrorIs(t, err2, tsdb.ErrPreviousShardFail{})
+			require.EqualError(t, err2, fmt.Errorf("not attempting to open shard %d; opening shard previously failed with: %w", shId, expErr).Error())
+
+			// Check that bad shard list now has a bad shard in it
+			require.Len(t, s.Store.GetBadShardList(), 1)
+		}()
+	}
+}
+
+func TestStore_BadShardClearNoBadShards(t *testing.T) {
+	indexes := tsdb.RegisteredIndexes()
+	for _, idx := range indexes {
+		func() {
+			s := MustOpenStore(t, idx)
+			defer require.NoErrorf(t, s.Close(), "closing store with index type: %s", idx)
+
+			sh := tsdb.NewTempShard(idx)
+			err := s.OpenShard(sh.Shard, false)
+			require.NoError(t, err, "opening temp shard")
+			defer require.NoError(t, sh.Close(), "closing temporary shard")
```

Comment:
As above, when is `sh.Close` called?

Reply:
Maybe a `defer` is not necessary here, since the shard is never used after opening.